tests/unit/test_ntm.py

```diff
@@ -0,0 +1,288 @@
+    sms = Mock(name='sagemaker_session', boto_session=boto_mock)
+    sms.boto_region_name = REGION
+    sms.default_bucket = Mock(name='default_bucket', return_value=BUCKET_NAME)
+    sms.sagemaker_client.describe_training_job = Mock(name='describe_training_job',
+                                                      return_value=DESCRIBE_TRAINING_JOB_RESULT)
+
+    return sms
+
+
+def test_init_required_positional(sagemaker_session):
+    ntm = NTM(ROLE, TRAIN_INSTANCE_COUNT, TRAIN_INSTANCE_TYPE, NUM_TOPICS, sagemaker_session=sagemaker_session)
+    assert ntm.role == ROLE
+    assert ntm.train_instance_count == TRAIN_INSTANCE_COUNT
+    assert ntm.train_instance_type == TRAIN_INSTANCE_TYPE
+    assert ntm.num_topics == NUM_TOPICS
+
+
+def test_init_required_named(sagemaker_session):
+    ntm = NTM(sagemaker_session=sagemaker_session, **ALL_REQ_ARGS)
+
+    assert ntm.role == COMMON_TRAIN_ARGS['role']
+    assert ntm.train_instance_count == TRAIN_INSTANCE_COUNT
+    assert ntm.train_instance_type == COMMON_TRAIN_ARGS['train_instance_type']
+    assert ntm.num_topics == ALL_REQ_ARGS['num_topics']
+
+
+def test_all_hyperparameters(sagemaker_session):
+    ntm = NTM(sagemaker_session=sagemaker_session,
+              encoder_layers=[1, 2, 'auto'], epochs=3, encoder_layers_activation='tanh', optimizer='sgd',
+              tolerance=0.05, num_patience_epochs=2, batch_norm=False, rescale_gradient=0.5, clip_gradient=0.5,
+              weight_decay=0.5, learning_rate=0.5, **ALL_REQ_ARGS)
+    assert ntm.hyperparameters() == dict(
+        num_topics=str(ALL_REQ_ARGS['num_topics']),
+        encoder_layers="[1, 2, 'auto']",
+        epochs='3',
+        encoder_layers_activation='tanh',
+        optimizer='sgd',
+        tolerance='0.05',
+        num_patience_epochs='2',
+        batch_norm='False',
+        rescale_gradient='0.5',
+        clip_gradient='0.5',
+        weight_decay='0.5',
+        learning_rate='0.5'
+    )
+
+
+def test_image(sagemaker_session):
+    ntm = NTM(sagemaker_session=sagemaker_session, **ALL_REQ_ARGS)
+    assert ntm.train_image() == registry(REGION, "ntm") + '/ntm:1'
+
+
+def test_num_topics_validation_fail_type(sagemaker_session):
+    with pytest.raises(ValueError):
+        NTM(num_topics='other', sagemaker_session=sagemaker_session, **COMMON_TRAIN_ARGS)
+
+
+def test_num_topics_validation_fail_value(sagemaker_session):
+    with pytest.raises(ValueError):
+        NTM(num_topics=0, sagemaker_session=sagemaker_session, **COMMON_TRAIN_ARGS)
+
+
+def test_encoder_layers_validation_fail_type(sagemaker_session):
+    with pytest.raises(TypeError):
+        NTM(encoder_layers=0, sagemaker_session=sagemaker_session, **ALL_REQ_ARGS)
+
+
+def test_epochs_validation_fail_type(sagemaker_session):
+    with pytest.raises(ValueError):
+        NTM(epochs='other', sagemaker_session=sagemaker_session, **ALL_REQ_ARGS)
+
+
+def test_epochs_validation_fail_value(sagemaker_session):
```

Comment:
Since the validation checks both min and max it would be great if we had both conditions checked for these HPs.

Reply:
Updated.

Now all hyper-parameters with a range will be validated for both lower and upper limit.